### DISCUSSION TITLE
Manim and Math Inspector

### DISCUSSION THREAD

**Discussion by MathInspector** [3b1b/manim#1411] [General]
Hi all!

First of all, a little backstory.  I am a HUGE fan of @3b1b and he inspired me to start my own math education youtube channel.  That was when I first found out about manim.   The problem I had was manim doesn't have a beautiful graphical user interface with block coding that made everything ridiculously easy to create production quality animations.  I got totally distracted by thinking about this and ended up creating a project whose vision is very closely aligned with manim, but which is actually at such a high level of generality that it is capable of mapping the entire python programming language into a block coding environment.

It's free and open source (gpl-gnu-v3) and is named math inspector (because it's like web inspector except for math).  I wanted to reach out to the community to find out if anyone would be interested in joining forces.

https://github.com/MathInspector/MathInspector

Wouldn't it be really cool to find a way to combine the visual quality of manim with a modern visual programming environment like math inspector?

- Matt

**Comment by huguesdevimeux** [comment]
Hi there, [manimCommunity](https://github.com/ManimCommunity/manim) developer here.

The project looks from my eyes very, very promising and interesting. From my personal opinion, I'd love to see that working with manim. 
I don't understand what you would need from manim, as as I saw from the videos of MathInspector there is already a wroking rendering process. 

Could you clarify what would be your plan? 

Hugues.

**Reply by MathInspector** [reply]
Thank you for the kind words @huguesdevimeux !

The reason I created math inspector was because I wanted to be part of the amazing revolution that is taking place in mathematics education on youtube. I wanted to take what I am good at and give back to the community.

I wanted to be able to pause 3b1b video's and use a modern block coding system that is standard in audio/visual editor tools and play around with the animations by dragging sliders and doing numerical experiments. I wanted to create a tool that could create a new kind of algorithmic, coding based problem sets, for a pure mathematics course.

I don't know the best way to work together, but have some initial ideas. The reason I am reaching out is to find out if there is a way to join forces and bring math software into the next generation, and create the next paradigm of what math software is.

> I don't understand what you would need from manim, as as I saw from the videos of MathInspector there is already a wroking rendering process.
Could you clarify what would be your plan?

Great question. It might be the case that mathinspector already works for manim, but since I have never gotten manim to work and am doing 1000 things I don't have time to check this right now.

One of the main reasons I wanted to reach out was to find out if it already works or not, and if it does, what does the manim community want to change about mathinspector to make it the ideal environment for using manim?

Here is how you could do it in mathinspector.

- Open math inspector
- select Add File from the project menu and choose an existing manim project
- all of the items will appear in the node editor and the output should be displayed in whatever window it is supposed to
- if the output doesnt show up, this is a bug in mathinspector, please open an issue and I'll try to fix it

The main difference between mathinspector and manims visual output is mathinspector allows you to interact with the plot while it's animating, using the techniques of modern game engines. Would this be a desirable feature for manim?

Down the line, I'd like to figure out the visual magic that makes manim tick, and have math inspectors plotting system be just as beautiful. In other words, turn manim into a video game, if that makes sense =)

**Reply by huguesdevimeux** [reply]
I see, but I still don't actually understand what is in manim that isn't actually in MathInspector. I mean, I saw that MathInspector has already a working rendering process for math animations, so why bother using manim ? Manim is basically what you implemented, no? Or there are missing features that you would like to use from manim?  

> The main difference between mathinspector and manims visual output is mathinspector allows you to interact with the plot while it's animating, using the techniques of modern game engines. Would this be a desirable feature for manim?

And yes, of course this is! ManimCommunity is moving towards a new rendering engine (OpenGL), so we will get closer to this at some point.

**Reply by MathInspector** [reply]
> Manim is basically what you implemented, no? 

That's true.  But I actually think the question you haven't asked is, why bother using math inspector for creating animations with Manim?

> Or there are missing features that you would like to use from manim?

Yes, this very much!  I guess my idea was something like, people who currently use Manim might find mathinspector a useful environment for working in.  If it is useful, maybe math inspector could add features to support Manim users better, and then eventually see where it goes down the line in terms of having a single render pipeline.

> I still don't actually understand what is in manim that isn't actually in MathInspector

Math inspector is a complete python environment, there is nothing "in python" that "isn't actually in MathInspector".  What math inspector does that standard python does not is create a visual programming interface on top of any existing python project without changing anything about the underlying code, including Manim.

> why bother using manim

From a graphic design perspective, imo, Manim's animations are in a class above what math inspectors look like.

> yes, of course this is! ManimCommunity is moving towards a new rendering engine (OpenGL)

My hope is: people won't think to themselves "should I use Manim or mathinspector for making animations" as if they are two competing products.  Part of the reason I created this discussion topic was to openly work through this kind of concern.  The visions that inspired Manim and mathinspector are very closely aligned, and that's why I wanted to open a discussion topic about finding out if there is a way to join forces.

**Reply by huguesdevimeux** [reply]
First of all, maybe you should join the [manim Community discord server](https://discord.gg/mMRrZQW). Much more devs and people interested in the project are there, so I'm sure the discussions will be more in-depth. 

> From a graphic design perspective, imo, Manim's animations are in a class above what math inspectors look like.

So, if I sum up, your plan would make the the logic of math inspector (the block-coding system) use manim instead of its own rendering process?  
If so, I think that's a very good idea! Others and I would happy to help with that. 

And, I just tested mathinspector, and I was just stunned. This is really incredible, very good job!

**Reply by MathInspector** [reply]
> I just tested mathinspector, and I was just stunned. This is really incredible, very good job!

Thank you!!!! =)

> So, if I sum up, your plan would make the the logic of math inspector (the block-coding system) use manim instead of its own rendering process?

Basically, yes, this is the plan I had in mind.  On a technical level, there are lots of reasons why that is a really big job, and my idea was to break it into steps, where the last step would be to use manim instead of the current render process.  Initially I thought it would be a good idea to do some experiments and identify the places where there seems to be a lot of potential but things in math inspector need fixing up to make it even easier to use.

If folks in the manim community finds mathinspector useful for working with manim code, then that would be step 0.  Step 1 would be learning more and understanding better the kind of UI changes that would it even more helpful, including adding more built-in feature support in math inspector for things that make it better for using manim.

> If so, I think that's a very good idea! Others and I would happy to help with that.

Awesome!  I just joined the discord.(@calhoun137)  imo the best way to proceed would be to not rush things, and there are many aspects of manim in terms of performance that might complicate the process of turning it into an interactive video game.  For math inspector I had to be *very very careful* all the time with the data and use all my best tricks to optimize how it all works, and I have absolutely no idea how manim works under the hood.  That's why at this time, many parts of the process of switching over to a manim render process are unclear.  

Just on a practical level, it might make sense to have more than a single render pipelines in math inspector, at least temporarily, one for heavy duty animations using manim as it currently exists, the other for doing the interactive plotting.  I use math inspector for my own video production work and would love to also use manim for the really advanced animations that currently I can't get.

**Reply by huguesdevimeux** [reply]
Nice. 
But once again, Manim is not designed yet to be interactive, so it's probably be nearly impossible to do something interactive (without modifying the source code of manim). 

Feel free if you need support or explanations on manim rendering process, the code is kind of though to understand.

**Reply by MathInspector** [reply]
Thanks, will do.

imo there is a strong use-case for creating animations that are very computationally expensive, and which are *not* interactive because it takes too long to render animations of this type.

If I have time I will try running basic manim animations from mathinspector, and perhaps will put up a repo with some working code.

I think this is a wonderful start =)

Cheers!
